### ISSUE TITLE
cgroups hook problem with hyperthreading

### ISSUE THREAD

**Issue by drtoss** [openpbs/openpbs#1817]
If the pbs_cgroups hook is enabled on a python3 PBS and vnode_per_numa_node, use_hyperthreads, and ncpus_are_cores are all set to true, and the cpuset cgroup is enabled, then you get an error when a float is used to index a list.

The problem is the python3 decision that division of integers should produce a float.

A fix, partially tested, is to use floor division:

```
--- /home7/dtalcott/work/drtaltc/src/hooks/cgroups/pbs_cgroups.PY       2020-06-08 12:02:08.930018109 -0700
+++ nas_cgroups.py      2020-06-09 08:26:44.430456554 -0700
@@ -1853,7 +1850,7 @@
                 if 'ht' in cpuinfo['cpu'][0]['flags']:
                     cpuinfo['hyperthreads_per_core'] = \
                         (cpuinfo['cpu'][0]['siblings']
-                         / cpuinfo['cpu'][0]['cpu cores'])
+                         // cpuinfo['cpu'][0]['cpu cores'])
                     # Map hyperthreads to physical cores
                     if cpuinfo['hyperthreads_per_core'] > 1:
                         pbs.logmsg(pbs.EVENT_DEBUG4,
@@ -1888,7 +1885,7 @@
         except Exception:
             pbs.logmsg(pbs.EVENT_DEBUG, '%s: Hyperthreading check failed' %
                        caller_name())
-        cpuinfo['physical_cpus'] = cpuinfo['logical_cpus'] / \
+        cpuinfo['physical_cpus'] = cpuinfo['logical_cpus'] // \
             cpuinfo['hyperthreads_per_core']
         pbs.logmsg(pbs.EVENT_DEBUG4, '%s returning: %s' %
                    (caller_name(), cpuinfo))
@@ -2247,11 +2244,11 @@
                     if not self.cfg['use_hyperthreads']:
                         # Do not treat a hyperthread as a core when
                         # use_hyperthreads is false.
-                        threads /= self.cpuinfo['hyperthreads_per_core']
+                        threads //= self.cpuinfo['hyperthreads_per_core']
                     elif self.cfg['ncpus_are_cores']:
                         # use_hyperthreads and ncpus_are_cores are both true,
                         # advertise only one thread per core
-                        threads /= self.cpuinfo['hyperthreads_per_core']
+                        threads //= self.cpuinfo['hyperthreads_per_core']
                     if vnodes:
                         # set the value on the host to 0
                         host_resc_avail['ncpus'] = 0

```

**Comment by alexis-cousein** [comment]
Could you post the error that is displayed in the MoM log?

**Comment by alexis-cousein** [comment]
The _type_ of a floor division is, of course, not guaranteed to be an integer. So I'd be more inclined to also _cast_ it to an integer.

**Comment by alexis-cousein** [comment]
OK. Error found:

06/09/2020 13:54:39.596226;0080;pbs_python;Hook;pbs_python;['Traceback (most recent call last):', '  File "<embedded code object>", line 5538, in main', '  File "<embedded code object>", line 968, in invoke_handler', '  File "<embedded code object>", line 1017, in _execjob_begin_handler', '  File "<embedded code object>", line 4569, in configure_job', '  File "<embedded code object>", line 3940, in assign_job', '  File "<embedded code object>", line 3778, in _assign_resources', 'TypeError: slice indices must be integers or None or have an __index__ method']

06/09/2020 13:54:39.596634;0001;pbs_python;Hook;pbs_python;Unexpected error in pbs_cgroups handling execjob_begin event for job 1.testdev-07-r7 (system hold set): TypeError ('slice indices must be integers or None or have an __index__ method',)

06/09/2020 13:54:39.596919;0100;pbs_python;Hook;pbs_python;Hook ended: pbs_cgroups, job ID 1.testdev-07-r7, event_type 64 (elapsed time: 0.0987)

06/09/2020 13:54:39.614075;0100;pbs_mom;Hook;pbs_cgroups;execjob_begin request rejected by 'pbs_cgroups'

06/09/2020 13:54:39.614132;0008;pbs_mom;Job;1.testdev-07-r7;Unexpected error in pbs_cgroups handling execjob_begin event for job 1.testdev-07-r7 (system hold set)

**Comment by alexis-cousein** [comment]
And yes, the floatiness propagated to "needed", which gave us trouble here:

            assigned['cpuset.cpus'] += corelist[:needed]

**Comment by drtoss** [comment]
> The _type_ of a floor division is, of course, not guaranteed to be an integer, even though the value is:
> 
> alexis@host:~> python3
> Python 3.4.6 (default, Mar 22 2017, 12:26:13) [GCC] on linux
> Type "help", "copyright", "credits" or "license" for more information.
> 
> > > > 6.0 // 2.0
> > > > 3.0
> 
> So I'd be more inclined to also _cast_ it to an integer. Yes, it "should" be unnecessary if the arguments are both integers, but as we can see, floaty integers can propagate quite rapidly, and casting to an integer will firewall that.

Agreed. That would be a better solution.

**Comment by drtoss** [comment]
Here is a new diff, with int casts added. Partially tested.

```
--- /home7/dtalcott/work/drtaltc/src/hooks/cgroups/pbs_cgroups.PY       2020-06-08 12:02:08.930018109 -0700
+++ nas_cgroups.py      2020-06-09 15:47:34.160261173 -0700
@@ -1852,8 +1849,8 @@
                     or 'AuthenticAMD' in cpuinfo['cpu'][0]['vendor_id']):
                 if 'ht' in cpuinfo['cpu'][0]['flags']:
                     cpuinfo['hyperthreads_per_core'] = \
-                        (cpuinfo['cpu'][0]['siblings']
-                         / cpuinfo['cpu'][0]['cpu cores'])
+                        int(cpuinfo['cpu'][0]['siblings']
+                         // cpuinfo['cpu'][0]['cpu cores'])
                     # Map hyperthreads to physical cores
                     if cpuinfo['hyperthreads_per_core'] > 1:
                         pbs.logmsg(pbs.EVENT_DEBUG4,
@@ -1888,8 +1885,8 @@
         except Exception:
             pbs.logmsg(pbs.EVENT_DEBUG, '%s: Hyperthreading check failed' %
                        caller_name())
-        cpuinfo['physical_cpus'] = cpuinfo['logical_cpus'] / \
-            cpuinfo['hyperthreads_per_core']
+        cpuinfo['physical_cpus'] = int(cpuinfo['logical_cpus'] //
+            cpuinfo['hyperthreads_per_core'])
         pbs.logmsg(pbs.EVENT_DEBUG4, '%s returning: %s' %
                    (caller_name(), cpuinfo))
         return cpuinfo
@@ -2247,11 +2244,11 @@
                     if not self.cfg['use_hyperthreads']:
                         # Do not treat a hyperthread as a core when
                         # use_hyperthreads is false.
-                        threads /= self.cpuinfo['hyperthreads_per_core']
+                        threads //= self.cpuinfo['hyperthreads_per_core']
                     elif self.cfg['ncpus_are_cores']:
                         # use_hyperthreads and ncpus_are_cores are both true,
                         # advertise only one thread per core
-                        threads /= self.cpuinfo['hyperthreads_per_core']
+                        threads //= self.cpuinfo['hyperthreads_per_core']
                     if vnodes:
                         # set the value on the host to 0
                         host_resc_avail['ncpus'] = 0

```

**Comment by vstumpf** [comment]
Fixed in 6381a52
